wasm/derive/src/validator/validate.rs

```diff
@@ -0,0 +1,207 @@
+//! Module with [`derive_validate`](crate::derive_validate) macro implementation
+
+use proc_macro2::Span;
+use syn::{Attribute, Ident, Path, Type};
+
+use super::*;
+
+/// [`derive_validate`](crate::derive_validate()) macro implementation
+pub fn impl_derive_validate(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let ident = input.ident;
+
+    let (validate_grant_impl, validate_revoke_impl) = gen_validate_impls(&input.attrs);
+
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+
+    quote! {
+        impl #impl_generics ::iroha_wasm::validator::traits::Validate for #ident #ty_generics
+        #where_clause
+        {
+            #validate_grant_impl
+            #validate_revoke_impl
+        }
+    }
+    .into()
+}
+
+/// Enum representing possible attributes.
+#[derive(Clone)]
+enum ValidateAttribute {
+    /// Represents just `validate` attribute.
+    General(Type),
+    /// Represents `validate_grant` and `validate_revoke` attributes together.
+    Separate {
+        grant_condition: Type,
+        revoke_condition: Type,
+    },
+}
+
+#[allow(clippy::fallible_impl_from)]
+impl<'attr, A: IntoIterator<Item = &'attr Attribute>> From<A> for ValidateAttribute {
```

Comment:
Even if the compiler doesn't yet, this behaviour doesn't follow the principle of least surprise. Why wasn't `TryFrom` considered instead?

Reply:
> Does compiler depend on this assumption and is it possible thatpanicking will result in UB?

I don't think so, it would be very strange. We shouldn't be able to break compiler assumptions without `unsafe` code.

> Why wasn't TryFrom considered instead?

Because it's a macro implementation detail. Macro implementation is split between different functions. If every function would return `Result` and then main macro function would check if some of them has returned an `Err` and only then panic that would be too much of useless code, because it's just a macro which should panic when an error happened. We can focus on happy path.

Note: there is #3234 issue

Reply:
> I don't think so, it would be very strange. We shouldn't be able to break compiler assumptions without unsafe code.

you are correct as far as the safe code is concerned, but there can be some unsafe code relying on this guarantee. 

* Example -  panicking across FFI is UB
Say you wanted to use `From::from` in an FFI function(extern function declarations are always unsafe) and thought you don't have to catch any panic because none of the functions you use can panic. If there is a code path where your `From::from` can panic your code is now unsound

Reply:
You shouldn't implement `From` if it's possible it will panic. I grant you that implementing `TryFrom` would be rather tedious for a macro functionality. Rather you can just make a custom freestanding function equivalent of `From::from` that will do the exact same work but can panic

Reply:
Good, will do